gptme/tools/__init__.py

```diff
@@ -36,84 +27,90 @@
     "ToolSpec",
     "ToolUse",
     "ToolFormat",
+    "Parameter",
+    "ConfirmFunc",
     # functions
-    "execute_msg",
     "get_tool_format",
     "set_tool_format",
-    # files
-    "read_tool",
-    "tool_append",
-    "tool_save",
-    "patch_tool",
-    # code
-    "shell_tool",
-    "python_tool",
-    "gh_tool",
-    # vision and computer use
-    "vision_tool",
-    "screenshot_tool",
-    "computer_tool",
-    # misc
-    "chats_tool",
-    "rag_tool",
-    "subagent_tool",
-    "tmux_tool",
-    "browser_tool",
-    "youtube_tool",
 ]
 
-loaded_tools: list[ToolSpec] = []
+from .save import tool_save
 
-# Tools that are disabled by default, unless explicitly enabled
-# TODO: find a better way to handle this
-tools_default_disabled = [
-    "computer",
-    "subagent",
-]
+_loaded_tools: list[ToolSpec] = []
+_available_tools: list[ToolSpec] | None = None
+
+
+def _discover_tools(package_names):
+    """Discover tools in a package or module, given the package name as a string."""
+    tools = []
+    for package_name in package_names:
+        try:
+            # Dynamically import the package or module
+            package = importlib.import_module(package_name)
+        except ModuleNotFoundError:
+            logger.warning("Module or package %s not found", package_name)
+            continue
+
+        # Check if it's a package or a module
+        if hasattr(package, "__path__"):  # It's a package
+            # Iterate over modules in the package
+            for _, module_name, _ in pkgutil.iter_modules(package.__path__):
+                full_module_name = f"{package_name}.{module_name}"
+                try:
+                    module = importlib.import_module(full_module_name)
+                except ModuleNotFoundError:
+                    logger.warning("Missing dependency for module %s", full_module_name)
+                    continue
+
+                # Find instances of ToolSpec in the module
+                for _, obj in inspect.getmembers(
+                    module, lambda c: isinstance(c, ToolSpec)
+                ):
+                    tools.append(obj)
+        else:  # It's a single module
+            # Find instances of ToolSpec in the module
+            for _, obj in inspect.getmembers(
+                package, lambda c: isinstance(c, ToolSpec)
+            ):
+                tools.append(obj)
```

Comment:
Can be rewritten something like:

```suggestion
        # Check if it's a package or a module
        if hasattr(package, "__path__"):  # It's a package
            # Iterate over modules in the package
            for _, module_name, _ in pkgutil.iter_modules(package.__path__):
                full_module_name = f"{package_name}.{module_name}"
                try:
                    module = importlib.import_module(full_module_name)
                except ModuleNotFoundError:
                    logger.warning("Missing dependency for module %s", full_module_name)
                    continue
            else:
                raise ValueError("no module in package")
        else:  # It's a single module
             module = package
            
        # Find instances of ToolSpec in the module
        for _, obj in inspect.getmembers(
            package, lambda c: isinstance(c, ToolSpec)
        ):
            tools.append(obj)
```

Not sure about the edge-case where a package has no modules.

Reply:
I think this version doesn't work when a package contains multiple modules but I guess you wanted to avoid the code duplication so I refactored the method, let me know what you think.